standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -6672,12 +6674,26 @@ public GetOpenTxnsInfoResponse get_open_txns_info() throws TException {
 
     @Override
     public OpenTxnsResponse open_txns(OpenTxnRequest rqst) throws TException {
-      return getTxnHandler().openTxns(rqst);
+      OpenTxnsResponse response = getTxnHandler().openTxns(rqst);
+      List<Long> txnIds = response.getTxn_ids();
+      if (!listeners.isEmpty()) {
+        MetaStoreListenerNotifier.notifyEvent(listeners, EventType.OPEN_TXN,
+            new OpenTxnEvent(txnIds.iterator(), this));
+      }
+      if (!transactionalListeners.isEmpty()) {
+        MetaStoreListenerNotifier.notifyEvent(transactionalListeners, EventType.OPEN_TXN,
+            new OpenTxnEvent(txnIds.iterator(), this));
```

Comment:
Noticed txnIds is passed like List->Iter->NewList->Iter->NewList to several methods. This model have several copies. Instead, can we pass the same list (txnIds) all the way down to build openTxnMessage?

Reply:
Fixed